File: 1052.爱生气的书店老板.ts
/*
 * @lc app=leetcode.cn id=1052 lang=typescript
 *
 * [1052] 爱生气的书店老板
 */
// n = 3
// [1,0,1,2,1,1,7,5]
// [0,1,0,1,0,1,0,1]
// @lc code=start
function maxSatisfied(customers: number[], grumpy: number[], minutes: number): number {
    let sum0 = 0
    for (let i = 0; i < grumpy.length; i++) {
        const element = grumpy[i];
        if (element === 0) {
            sum0 += customers[i]
        }
    }
    let res = 0
    let sum1 = 0
    for (let i = 0; i < minutes; i++) {
        const element = grumpy[i];
        if (element === 1) {
            sum1 += customers[i]
        }
    }
    res = sum1
    for (let i = minutes; i < grumpy.length; i++) {
        const element = grumpy[i];
        if (element === 1) {
            sum1 += customers[i]
        }
        if (grumpy[i - minutes] === 1) {
            sum1 -= customers[i - minutes]
        }
        res = Math.max(res, sum1)
    }
    return sum0 + res
};
// @lc code=end

File: 1052.爱生气的书店老板.go
/*
 * @lc app=leetcode.cn id=1052 lang=golang
 *
 * [1052] 爱生气的书店老板
 */

// @lc code=start
func maxSatisfied(customers []int, grumpy []int, minutes int) int {
    s := [2]int{}
	res := 0
	for k, v := range customers {
		s[grumpy[k]] += v
		if k < minutes - 1 {
			continue
		}
		res = max(res, s[1])
		if grumpy[k-minutes +1] == 1 {
			s[1] -= customers[k-minutes+1]
		}
	}
	return res + s[0]
}
// @lc code=end

